components/resource_control/src/worker.rs

```diff
@@ -0,0 +1,425 @@
+                self.process_stat.cpu_usage().map(|u| ResourceUsageStats {
+                    // cpu is measured in us.
+                    total_quota: total_quota * 1_000_000.0,
+                    current_used: u * 1_000_000.0,
+                })
+            }
+            ResourceType::Io => {
+                let mut stats = ResourceUsageStats {
+                    total_quota: self.io_bandwidth as f64,
+                    current_used: 0.0,
+                };
+                let now = Instant::now_coarse();
+                let dur = now.saturating_duration_since(self.prev_io_ts).as_secs_f64();
+                if dur < 0.1 {
+                    return Ok(stats);
+                }
+                let new_io_stats = fetch_io_bytes();
+                let total_io_used = self
+                    .prev_io_stats
+                    .iter()
+                    .zip(new_io_stats.iter())
+                    .map(|(s, new_s)| {
+                        let delta = *new_s - *s;
+                        delta.read + delta.write
+                    })
+                    .sum::<u64>();
+                self.prev_io_stats = new_io_stats;
+                self.prev_io_ts = now;
+
+                stats.current_used = total_io_used as f64 / dur;
+                Ok(stats)
+            }
+        }
+    }
+}
+
+pub struct GroupQuotaAdjustWorker<R> {
+    prev_stats_by_group: [HashMap<String, GroupStatistics>; ResourceType::COUNT],
+    last_adjust_time: Instant,
+    resource_ctl: Arc<ResourceGroupManager>,
+    is_last_time_low_load: [bool; ResourceType::COUNT],
+    resource_quota_getter: R,
+}
+
+impl GroupQuotaAdjustWorker<SysQuotaGetter> {
+    pub fn new(resource_ctl: Arc<ResourceGroupManager>, io_bandwidth: u64) -> Self {
+        let resource_quota_getter = SysQuotaGetter {
+            process_stat: ProcessStat::cur_proc_stat().unwrap(),
+            prev_io_stats: [IoBytes::default(); IoType::COUNT],
+            prev_io_ts: Instant::now_coarse(),
+            io_bandwidth,
+        };
+        Self::with_quota_getter(resource_ctl, resource_quota_getter)
+    }
+}
+
+impl<R: ResourceStatsProvider> GroupQuotaAdjustWorker<R> {
+    pub fn with_quota_getter(
+        resource_ctl: Arc<ResourceGroupManager>,
+        resource_quota_getter: R,
+    ) -> Self {
+        let prev_stats_by_group = array::from_fn(|_| HashMap::default());
+        Self {
+            prev_stats_by_group,
+            last_adjust_time: Instant::now_coarse(),
+            resource_ctl,
+            resource_quota_getter,
+            is_last_time_low_load: array::from_fn(|_| false),
+        }
+    }
+
+    pub fn adjust_quota(&mut self) {
+        let now = Instant::now_coarse();
+        let dur_secs = now
+            .saturating_duration_since(self.last_adjust_time)
+            .as_secs_f64();
+        self.last_adjust_time = now;
+        if dur_secs < 1.0 {
+            return;
+        }
```

Comment:
```suggestion
        if dur_secs < 1.0 {
            return;
        }
        self.last_adjust_time = now;
```

Reply:
maybe need to add a test to check for background tasks with the interval of less than 1 second.